{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "import networkx as nx\n",
      "import importlib\n",
      "import logging\n",
      "import functools\n",
      "from collections import OrderedDict\n",
      "\n",
      "import pbsmrtpipe.loader as pbl\n",
      "import pbsmrtpipe.chunked_models as pbc\n",
      "import pbsmrtpipe.dag as pbd\n",
      "import pbsmrtpipe.io as pbio\n",
      "import pbsmrtpipe.resource_types as pbrt\n",
      "import pbsmrtpipe.utils as pbu\n",
      "import pbsmrtpipe.mock\n",
      "\n",
      "import pbsmrtpipe.task_modules\n",
      "from pbsmrtpipe.resource_types import ReferenceEntryType, InputXml, MovieFofn, BaseFileType\n",
      "from pbsmrtpipe.decos import validate_type\n",
      "from pbsmrtpipe.model import P_Module, MetaTask, MetaFile, FileNode, TaskNode\n",
      "from pbsmrtpipe.task_modules.core import ChunkedList, BaseGatherTask\n",
      "\n",
      "import pbsmrtpipe.toolbook as tb\n",
      "\n",
      "log = logging.getLogger()\n",
      "#pbu.setup_log(log, level=logging.DEBUG)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pb_module_names = ['preassembler_dagcon']\n",
      "\n",
      "task_klasses = pbl.load_tasks_from_pb_module_names(pb_module_names)\n",
      "abstract_g = pbd.build_abstract_graph_from_task_classes(task_klasses)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pbd.find_required_input_files(abstract_g)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 3,
       "text": [
        "[<MetaFile id 17 file PacBioToCASemFrgLoaded>,\n",
        " <MetaFile id 23 file PacBioToCASem>,\n",
        " <MetaFile id 1 file FilteredSubreadsFasta>]"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tb.display_networkx_graph(abstract_g)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "svg": [
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
        "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
        " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
        "<!-- Generated by graphviz version 2.34.0 (20130907.0228)\n",
        " -->\n",
        "<!-- Title: G Pages: 1 -->\n",
        "<svg width=\"1511pt\" height=\"908pt\"\n",
        " viewBox=\"0.00 0.00 1511.00 908.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
        "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 904)\">\n",
        "<title>G</title>\n",
        "<polygon fill=\"white\" stroke=\"white\" points=\"-4,4 -4,-904 1507,-904 1507,4 -4,4\"/>\n",
        "<!-- MetaFile id 18 file RunCeleraAssemblerSpec -->\n",
        "<g id=\"node1\" class=\"node\"><title>MetaFile id 18 file RunCeleraAssemblerSpec</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"492.25,-324 221.75,-324 221.75,-288 492.25,-288 492.25,-324\"/>\n",
        "<text text-anchor=\"middle\" x=\"357\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 18 file RunCeleraAssemblerSpec</text>\n",
        "</g>\n",
        "<!-- MetaTask id 16 task RunHgapCaTask -->\n",
        "<g id=\"node6\" class=\"node\"><title>MetaTask id 16 task RunHgapCaTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"505\" cy=\"-234\" rx=\"149.23\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"505\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 16 task RunHgapCaTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 18 file RunCeleraAssemblerSpec&#45;&gt;MetaTask id 16 task RunHgapCaTask -->\n",
        "<g id=\"edge1\" class=\"edge\"><title>MetaFile id 18 file RunCeleraAssemblerSpec&#45;&gt;MetaTask id 16 task RunHgapCaTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M393.205,-287.876C413.556,-278.251 439.12,-266.16 460.664,-255.97\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"462.391,-259.025 469.934,-251.585 459.398,-252.697 462.391,-259.025\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 4 file BlasrM4 -->\n",
        "<g id=\"node2\" class=\"node\"><title>MetaFile id 4 file BlasrM4</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"903,-612 737,-612 737,-576 903,-576 903,-612\"/>\n",
        "<text text-anchor=\"middle\" x=\"820\" y=\"-590.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 4 file BlasrM4</text>\n",
        "</g>\n",
        "<!-- MetaTask id 6 task HgapFilterM4Task -->\n",
        "<g id=\"node8\" class=\"node\"><title>MetaTask id 6 task HgapFilterM4Task</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"663\" cy=\"-522\" rx=\"152.13\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"663\" y=\"-518.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 6 task HgapFilterM4Task</text>\n",
        "</g>\n",
        "<!-- MetaFile id 4 file BlasrM4&#45;&gt;MetaTask id 6 task HgapFilterM4Task -->\n",
        "<g id=\"edge3\" class=\"edge\"><title>MetaFile id 4 file BlasrM4&#45;&gt;MetaTask id 6 task HgapFilterM4Task</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M781.593,-575.876C759.732,-566.129 732.199,-553.853 709.166,-543.584\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"710.487,-540.341 699.929,-539.465 707.637,-546.734 710.487,-540.341\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 8 task HgapCorrectionTask -->\n",
        "<g id=\"node24\" class=\"node\"><title>MetaTask id 8 task HgapCorrectionTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"1017\" cy=\"-522\" rx=\"158.379\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"1017\" y=\"-518.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 8 task HgapCorrectionTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 4 file BlasrM4&#45;&gt;MetaTask id 8 task HgapCorrectionTask -->\n",
        "<g id=\"edge2\" class=\"edge\"><title>MetaFile id 4 file BlasrM4&#45;&gt;MetaTask id 8 task HgapCorrectionTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M867.94,-575.966C896.181,-565.931 932.031,-553.192 961.495,-542.723\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"962.971,-545.913 971.222,-539.266 960.627,-539.317 962.971,-545.913\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 25 file CorrectedReadsFrg -->\n",
        "<g id=\"node3\" class=\"node\"><title>MetaFile id 25 file CorrectedReadsFrg</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1048.25,-324 815.75,-324 815.75,-288 1048.25,-288 1048.25,-324\"/>\n",
        "<text text-anchor=\"middle\" x=\"932\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 25 file CorrectedReadsFrg</text>\n",
        "</g>\n",
        "<!-- MetaFile id 9 file CorrectedFasta -->\n",
        "<g id=\"node4\" class=\"node\"><title>MetaFile id 9 file CorrectedFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"952,-468 750,-468 750,-432 952,-432 952,-468\"/>\n",
        "<text text-anchor=\"middle\" x=\"851\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 9 file CorrectedFasta</text>\n",
        "</g>\n",
        "<!-- MetaTask id 22 task WriteRunCASpecTask -->\n",
        "<g id=\"node14\" class=\"node\"><title>MetaTask id 22 task WriteRunCASpecTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"361\" cy=\"-378\" rx=\"170.402\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"361\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 22 task WriteRunCASpecTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 9 file CorrectedFasta&#45;&gt;MetaTask id 22 task WriteRunCASpecTask -->\n",
        "<g id=\"edge4\" class=\"edge\"><title>MetaFile id 9 file CorrectedFasta&#45;&gt;MetaTask id 22 task WriteRunCASpecTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M749.978,-433.471C746.612,-432.972 743.279,-432.48 740,-432 649.406,-418.729 546.517,-404.387 470.825,-393.973\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"471.181,-390.489 460.797,-392.594 470.227,-397.424 471.181,-390.489\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 11 task PreAssemblerJsonReportTask -->\n",
        "<g id=\"node23\" class=\"node\"><title>MetaTask id 11 task PreAssemblerJsonReportTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"1307\" cy=\"-378\" rx=\"196.399\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"1307\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 11 task PreAssemblerJsonReportTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 9 file CorrectedFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask -->\n",
        "<g id=\"edge5\" class=\"edge\"><title>MetaFile id 9 file CorrectedFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M952.001,-433.385C955.034,-432.918 958.04,-432.455 961,-432 1041.43,-419.624 1132.39,-405.693 1200.86,-395.221\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1201.44,-398.673 1210.79,-393.701 1200.38,-391.754 1201.44,-398.673\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 10 file CorrectedFastq -->\n",
        "<g id=\"node5\" class=\"node\"><title>MetaFile id 10 file CorrectedFastq</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1179.25,-468 970.75,-468 970.75,-432 1179.25,-432 1179.25,-468\"/>\n",
        "<text text-anchor=\"middle\" x=\"1075\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 10 file CorrectedFastq</text>\n",
        "</g>\n",
        "<!-- MetaTask id 24 task GenerateFrgFileTask -->\n",
        "<g id=\"node15\" class=\"node\"><title>MetaTask id 24 task GenerateFrgFileTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"928\" cy=\"-378\" rx=\"164.153\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"928\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 24 task GenerateFrgFileTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 10 file CorrectedFastq&#45;&gt;MetaTask id 24 task GenerateFrgFileTask -->\n",
        "<g id=\"edge6\" class=\"edge\"><title>MetaFile id 10 file CorrectedFastq&#45;&gt;MetaTask id 24 task GenerateFrgFileTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1039.04,-431.876C1018.9,-422.286 993.621,-410.248 972.272,-400.082\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"973.615,-396.845 963.082,-395.706 970.606,-403.165 973.615,-396.845\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 19 file CeleraAssemblerAsm -->\n",
        "<g id=\"node10\" class=\"node\"><title>MetaFile id 19 file CeleraAssemblerAsm</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"246.25,-180 -0.25,-180 -0.25,-144 246.25,-144 246.25,-180\"/>\n",
        "<text text-anchor=\"middle\" x=\"123\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 19 file CeleraAssemblerAsm</text>\n",
        "</g>\n",
        "<!-- MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 19 file CeleraAssemblerAsm -->\n",
        "<g id=\"edge10\" class=\"edge\"><title>MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 19 file CeleraAssemblerAsm</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M426.775,-218.666C368.76,-208.035 289.119,-193.441 225.937,-181.863\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"226.411,-178.391 215.944,-180.032 225.149,-185.277 226.411,-178.391\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 20 file AssemblyDegFasta -->\n",
        "<g id=\"node13\" class=\"node\"><title>MetaFile id 20 file AssemblyDegFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"497.25,-180 264.75,-180 264.75,-144 497.25,-144 497.25,-180\"/>\n",
        "<text text-anchor=\"middle\" x=\"381\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 20 file AssemblyDegFasta</text>\n",
        "</g>\n",
        "<!-- MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 20 file AssemblyDegFasta -->\n",
        "<g id=\"edge8\" class=\"edge\"><title>MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 20 file AssemblyDegFasta</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M475.299,-216.233C458.825,-206.934 438.131,-195.251 420.328,-185.201\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"421.749,-181.984 411.32,-180.116 418.308,-188.08 421.749,-181.984\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 14 file AssemblyScfFasta -->\n",
        "<g id=\"node20\" class=\"node\"><title>MetaFile id 14 file AssemblyScfFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"744,-180 516,-180 516,-144 744,-144 744,-180\"/>\n",
        "<text text-anchor=\"middle\" x=\"630\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 14 file AssemblyScfFasta</text>\n",
        "</g>\n",
        "<!-- MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 14 file AssemblyScfFasta -->\n",
        "<g id=\"edge7\" class=\"edge\"><title>MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 14 file AssemblyScfFasta</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M534.941,-216.233C551.547,-206.934 572.409,-195.251 590.354,-185.201\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"592.42,-188.056 599.435,-180.116 589,-181.949 592.42,-188.056\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 21 file AssemblySingletonFasta -->\n",
        "<g id=\"node25\" class=\"node\"><title>MetaFile id 21 file AssemblySingletonFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1025.25,-180 762.75,-180 762.75,-144 1025.25,-144 1025.25,-180\"/>\n",
        "<text text-anchor=\"middle\" x=\"894\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 21 file AssemblySingletonFasta</text>\n",
        "</g>\n",
        "<!-- MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 21 file AssemblySingletonFasta -->\n",
        "<g id=\"edge9\" class=\"edge\"><title>MetaTask id 16 task RunHgapCaTask&#45;&gt;MetaFile id 21 file AssemblySingletonFasta</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M584.659,-218.666C643.736,-208.035 724.837,-193.441 789.177,-181.863\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"790.131,-185.247 799.353,-180.032 788.891,-178.358 790.131,-185.247\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 0 task FilterLongReadsByLengthTask -->\n",
        "<g id=\"node7\" class=\"node\"><title>MetaTask id 0 task FilterLongReadsByLengthTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"1043\" cy=\"-810\" rx=\"197.374\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"1043\" y=\"-806.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 0 task FilterLongReadsByLengthTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 2 file FilteredLongReadsFasta -->\n",
        "<g id=\"node11\" class=\"node\"><title>MetaFile id 2 file FilteredLongReadsFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1170,-756 916,-756 916,-720 1170,-720 1170,-756\"/>\n",
        "<text text-anchor=\"middle\" x=\"1043\" y=\"-734.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 2 file FilteredLongReadsFasta</text>\n",
        "</g>\n",
        "<!-- MetaTask id 0 task FilterLongReadsByLengthTask&#45;&gt;MetaFile id 2 file FilteredLongReadsFasta -->\n",
        "<g id=\"edge11\" class=\"edge\"><title>MetaTask id 0 task FilterLongReadsByLengthTask&#45;&gt;MetaFile id 2 file FilteredLongReadsFasta</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1043,-791.697C1043,-783.983 1043,-774.712 1043,-766.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1046.5,-766.104 1043,-756.104 1039.5,-766.104 1046.5,-766.104\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 7 file FilteredBlasrM4 -->\n",
        "<g id=\"node12\" class=\"node\"><title>MetaFile id 7 file FilteredBlasrM4</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"731.25,-468 522.75,-468 522.75,-432 731.25,-432 731.25,-468\"/>\n",
        "<text text-anchor=\"middle\" x=\"627\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 7 file FilteredBlasrM4</text>\n",
        "</g>\n",
        "<!-- MetaTask id 6 task HgapFilterM4Task&#45;&gt;MetaFile id 7 file FilteredBlasrM4 -->\n",
        "<g id=\"edge12\" class=\"edge\"><title>MetaTask id 6 task HgapFilterM4Task&#45;&gt;MetaFile id 7 file FilteredBlasrM4</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M654.101,-503.697C650.003,-495.728 645.051,-486.1 640.507,-477.264\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"643.483,-475.396 635.797,-468.104 637.258,-478.598 643.483,-475.396\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 1 file FilteredSubreadsFasta -->\n",
        "<g id=\"node9\" class=\"node\"><title>MetaFile id 1 file FilteredSubreadsFasta</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1276.25,-900 1035.75,-900 1035.75,-864 1276.25,-864 1276.25,-900\"/>\n",
        "<text text-anchor=\"middle\" x=\"1156\" y=\"-878.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 1 file FilteredSubreadsFasta</text>\n",
        "</g>\n",
        "<!-- MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 0 task FilterLongReadsByLengthTask -->\n",
        "<g id=\"edge16\" class=\"edge\"><title>MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 0 task FilterLongReadsByLengthTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1128.36,-863.876C1113.47,-854.655 1094.93,-843.172 1078.94,-833.265\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1080.7,-830.238 1070.36,-827.947 1077.02,-836.189 1080.7,-830.238\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 3 task HgapAlignForCorrectionTask -->\n",
        "<g id=\"node18\" class=\"node\"><title>MetaTask id 3 task HgapAlignForCorrectionTask</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"836\" cy=\"-666\" rx=\"192.075\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"836\" y=\"-662.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 3 task HgapAlignForCorrectionTask</text>\n",
        "</g>\n",
        "<!-- MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 3 task HgapAlignForCorrectionTask -->\n",
        "<g id=\"edge14\" class=\"edge\"><title>MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 3 task HgapAlignForCorrectionTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1035.41,-876.824C955.809,-871.236 861.631,-858.225 836,-828 804.409,-790.746 815.947,-729.332 826.453,-694.101\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"829.817,-695.072 829.492,-684.482 823.142,-692.963 829.817,-695.072\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask -->\n",
        "<g id=\"edge15\" class=\"edge\"><title>MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1214.57,-863.911C1234.54,-855.708 1255.68,-844.068 1271,-828 1300.45,-797.121 1307,-781.669 1307,-739 1307,-739 1307,-739 1307,-521 1307,-480.996 1307,-434.653 1307,-406.08\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1310.5,-406.049 1307,-396.049 1303.5,-406.049 1310.5,-406.049\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 8 task HgapCorrectionTask -->\n",
        "<g id=\"edge13\" class=\"edge\"><title>MetaFile id 1 file FilteredSubreadsFasta&#45;&gt;MetaTask id 8 task HgapCorrectionTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1202.79,-863.807C1219.66,-855.422 1237.4,-843.687 1249,-828 1273.11,-795.407 1269,-779.542 1269,-739 1269,-739 1269,-739 1269,-665 1269,-592.458 1188.76,-556.569 1118.84,-539.067\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1119.59,-535.648 1109.05,-536.717 1117.95,-542.455 1119.59,-535.648\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 2 file FilteredLongReadsFasta&#45;&gt;MetaTask id 3 task HgapAlignForCorrectionTask -->\n",
        "<g id=\"edge17\" class=\"edge\"><title>MetaFile id 2 file FilteredLongReadsFasta&#45;&gt;MetaTask id 3 task HgapAlignForCorrectionTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M992.627,-719.966C963.039,-709.96 925.503,-697.267 894.595,-686.815\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"895.406,-683.394 884.811,-683.506 893.163,-690.025 895.406,-683.394\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 2 file FilteredLongReadsFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask -->\n",
        "<g id=\"edge18\" class=\"edge\"><title>MetaFile id 2 file FilteredLongReadsFasta&#45;&gt;MetaTask id 11 task PreAssemblerJsonReportTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1097.49,-719.894C1150.47,-699.463 1223,-659.681 1223,-595 1223,-595 1223,-595 1223,-521 1223,-474.596 1256.64,-430.055 1281.34,-403.604\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1283.95,-405.948 1288.36,-396.315 1278.9,-401.094 1283.95,-405.948\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 22 task WriteRunCASpecTask&#45;&gt;MetaFile id 18 file RunCeleraAssemblerSpec -->\n",
        "<g id=\"edge19\" class=\"edge\"><title>MetaTask id 22 task WriteRunCASpecTask&#45;&gt;MetaFile id 18 file RunCeleraAssemblerSpec</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M360.011,-359.697C359.57,-351.983 359.041,-342.712 358.549,-334.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"362.042,-333.888 357.977,-324.104 355.054,-334.288 362.042,-333.888\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 24 task GenerateFrgFileTask&#45;&gt;MetaFile id 25 file CorrectedReadsFrg -->\n",
        "<g id=\"edge20\" class=\"edge\"><title>MetaTask id 24 task GenerateFrgFileTask&#45;&gt;MetaFile id 25 file CorrectedReadsFrg</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M928.989,-359.697C929.43,-351.983 929.959,-342.712 930.451,-334.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"933.946,-334.288 931.023,-324.104 926.958,-333.888 933.946,-334.288\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 5 file BlasrM4Fofn -->\n",
        "<g id=\"node16\" class=\"node\"><title>MetaFile id 5 file BlasrM4Fofn</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1113,-612 921,-612 921,-576 1113,-576 1113,-612\"/>\n",
        "<text text-anchor=\"middle\" x=\"1017\" y=\"-590.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 5 file BlasrM4Fofn</text>\n",
        "</g>\n",
        "<!-- MetaFile id 5 file BlasrM4Fofn&#45;&gt;MetaTask id 8 task HgapCorrectionTask -->\n",
        "<g id=\"edge21\" class=\"edge\"><title>MetaFile id 5 file BlasrM4Fofn&#45;&gt;MetaTask id 8 task HgapCorrectionTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1017,-575.697C1017,-567.983 1017,-558.712 1017,-550.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1020.5,-550.104 1017,-540.104 1013.5,-550.104 1020.5,-550.104\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 12 file PreAssemblerReport -->\n",
        "<g id=\"node17\" class=\"node\"><title>MetaFile id 12 file PreAssemblerReport</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"1427,-324 1187,-324 1187,-288 1427,-288 1427,-324\"/>\n",
        "<text text-anchor=\"middle\" x=\"1307\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 12 file PreAssemblerReport</text>\n",
        "</g>\n",
        "<!-- MetaTask id 3 task HgapAlignForCorrectionTask&#45;&gt;MetaFile id 4 file BlasrM4 -->\n",
        "<g id=\"edge23\" class=\"edge\"><title>MetaTask id 3 task HgapAlignForCorrectionTask&#45;&gt;MetaFile id 4 file BlasrM4</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M832.045,-647.697C830.282,-639.983 828.163,-630.712 826.197,-622.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"829.55,-621.073 823.91,-612.104 822.726,-622.633 829.55,-621.073\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 3 task HgapAlignForCorrectionTask&#45;&gt;MetaFile id 5 file BlasrM4Fofn -->\n",
        "<g id=\"edge22\" class=\"edge\"><title>MetaTask id 3 task HgapAlignForCorrectionTask&#45;&gt;MetaFile id 5 file BlasrM4Fofn</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M878.894,-648.411C904.126,-638.653 936.259,-626.226 963.18,-615.814\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"964.776,-618.95 972.84,-612.078 962.251,-612.421 964.776,-618.95\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 13 task RunHgapReferenceUploader -->\n",
        "<g id=\"node19\" class=\"node\"><title>MetaTask id 13 task RunHgapReferenceUploader</title>\n",
        "<ellipse fill=\"#dddddd\" stroke=\"black\" cx=\"630\" cy=\"-90\" rx=\"192.075\" ry=\"18\"/>\n",
        "<text text-anchor=\"middle\" x=\"630\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaTask id 13 task RunHgapReferenceUploader</text>\n",
        "</g>\n",
        "<!-- MetaFile id 15 file ReferenceEntryType -->\n",
        "<g id=\"node22\" class=\"node\"><title>MetaFile id 15 file ReferenceEntryType</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"750,-36 510,-36 510,-0 750,-0 750,-36\"/>\n",
        "<text text-anchor=\"middle\" x=\"630\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 15 file ReferenceEntryType</text>\n",
        "</g>\n",
        "<!-- MetaTask id 13 task RunHgapReferenceUploader&#45;&gt;MetaFile id 15 file ReferenceEntryType -->\n",
        "<g id=\"edge24\" class=\"edge\"><title>MetaTask id 13 task RunHgapReferenceUploader&#45;&gt;MetaFile id 15 file ReferenceEntryType</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M630,-71.6966C630,-63.9827 630,-54.7125 630,-46.1124\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"633.5,-46.1043 630,-36.1043 626.5,-46.1044 633.5,-46.1043\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 14 file AssemblyScfFasta&#45;&gt;MetaTask id 13 task RunHgapReferenceUploader -->\n",
        "<g id=\"edge25\" class=\"edge\"><title>MetaFile id 14 file AssemblyScfFasta&#45;&gt;MetaTask id 13 task RunHgapReferenceUploader</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M630,-143.697C630,-135.983 630,-126.712 630,-118.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"633.5,-118.104 630,-108.104 626.5,-118.104 633.5,-118.104\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 23 file PacBioToCASem -->\n",
        "<g id=\"node21\" class=\"node\"><title>MetaFile id 23 file PacBioToCASem</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"473,-468 249,-468 249,-432 473,-432 473,-468\"/>\n",
        "<text text-anchor=\"middle\" x=\"361\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 23 file PacBioToCASem</text>\n",
        "</g>\n",
        "<!-- MetaFile id 23 file PacBioToCASem&#45;&gt;MetaTask id 22 task WriteRunCASpecTask -->\n",
        "<g id=\"edge26\" class=\"edge\"><title>MetaFile id 23 file PacBioToCASem&#45;&gt;MetaTask id 22 task WriteRunCASpecTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M361,-431.697C361,-423.983 361,-414.712 361,-406.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"364.5,-406.104 361,-396.104 357.5,-406.104 364.5,-406.104\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 11 task PreAssemblerJsonReportTask&#45;&gt;MetaFile id 12 file PreAssemblerReport -->\n",
        "<g id=\"edge27\" class=\"edge\"><title>MetaTask id 11 task PreAssemblerJsonReportTask&#45;&gt;MetaFile id 12 file PreAssemblerReport</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1307,-359.697C1307,-351.983 1307,-342.712 1307,-334.112\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1310.5,-334.104 1307,-324.104 1303.5,-334.104 1310.5,-334.104\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 8 task HgapCorrectionTask&#45;&gt;MetaFile id 9 file CorrectedFasta -->\n",
        "<g id=\"edge28\" class=\"edge\"><title>MetaTask id 8 task HgapCorrectionTask&#45;&gt;MetaFile id 9 file CorrectedFasta</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M977.66,-504.411C954.723,-494.739 925.566,-482.444 901.011,-472.089\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"902.075,-468.739 891.5,-468.078 899.355,-475.189 902.075,-468.739\"/>\n",
        "</g>\n",
        "<!-- MetaTask id 8 task HgapCorrectionTask&#45;&gt;MetaFile id 10 file CorrectedFastq -->\n",
        "<g id=\"edge29\" class=\"edge\"><title>MetaTask id 8 task HgapCorrectionTask&#45;&gt;MetaFile id 10 file CorrectedFastq</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M1031.04,-504.055C1037.99,-495.668 1046.52,-485.376 1054.21,-476.096\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"1057,-478.209 1060.69,-468.275 1051.61,-473.742 1057,-478.209\"/>\n",
        "</g>\n",
        "<!-- MetaFile id 17 file PacBioToCASemFrgLoaded -->\n",
        "<g id=\"node26\" class=\"node\"><title>MetaFile id 17 file PacBioToCASemFrgLoaded</title>\n",
        "<polygon fill=\"none\" stroke=\"black\" points=\"797,-324 511,-324 511,-288 797,-288 797,-324\"/>\n",
        "<text text-anchor=\"middle\" x=\"654\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">MetaFile id 17 file PacBioToCASemFrgLoaded</text>\n",
        "</g>\n",
        "<!-- MetaFile id 17 file PacBioToCASemFrgLoaded&#45;&gt;MetaTask id 16 task RunHgapCaTask -->\n",
        "<g id=\"edge30\" class=\"edge\"><title>MetaFile id 17 file PacBioToCASemFrgLoaded&#45;&gt;MetaTask id 16 task RunHgapCaTask</title>\n",
        "<path fill=\"none\" stroke=\"black\" d=\"M617.55,-287.876C597.062,-278.251 571.325,-266.16 549.636,-255.97\"/>\n",
        "<polygon fill=\"black\" stroke=\"black\" points=\"550.842,-252.67 540.303,-251.585 547.865,-259.005 550.842,-252.67\"/>\n",
        "</g>\n",
        "</g>\n",
        "</svg>\n"
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "workflow_settings = pbsmrtpipe.mock.get_hgap2_workflow_settings()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print workflow_settings.summary()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "WorkflowSettings Summary\n",
        "           chunked : False\n",
        "       distributed : False\n",
        "         max_slots : 8\n",
        "       max_workers : 6\n",
        "           tmp_dir : /scratch\n",
        "             nproc : 4\n",
        "            job_id : 123456\n",
        "             debug : True\n",
        "    job_output_dir : /tmp/mock_job\n",
        "  ALGO_PARAMS_PATH : /Users/mkocher/MOCK_SEYMOUR_HOME/analysis/etc/algorithm_parameters/2013-09\n",
        "    INPUT_HDF5_EXT : bas.h5\n",
        "\n",
        "Module Parameters\n",
        "******************************\n",
        "module scope: p_mappingreports\n",
        "******************************\n",
        "              moduleName : P_MappingReports\n",
        "***********************\n",
        "module scope: p_mapping\n",
        "***********************\n",
        "                 maxHits : 10\n",
        "           maxDivergence : 30\n",
        "              moduleName : P_MappingReports\n",
        "            pulseMetrics : DeletionQV,IPD,InsertionQV,PulseWidth,QualityValue,MergeQV,SubstitutionQV,DeletionTag\n",
        "              moduleName : P_Mapping\n",
        "                  samBam : True\n",
        "    placeRepeatsRandomly : True\n",
        "           minAnchorSize : 12\n",
        "             description : BLASR maps reads to genomes by finding the highest scoring local alignment or set of local alignments between the read and the genome. The first set of alignments is found by querying an index of the reference genome, and then refining until only high scoring alignments are retained.  Additional pulse metrics are loaded into the resulting cmp.h5 file to enable downstream use of the Quiver algorithm.\n",
        "          loadPulsesOpts : bymetric\n",
        "                 gff2Bed : True\n",
        "            pbalign_opts : --seed=1 --minAccuracy=0.75 --minLength=50 --algorithmOptions=\"-useQuality\"\n",
        "*********************************\n",
        "module scope: p_referenceuploader\n",
        "*********************************\n",
        "              moduleName : P_ReferenceUploader\n",
        "         runUploaderHgap : True\n",
        "                sawriter : sawriter -blt 8 -welter\n",
        "             runUploader : False\n",
        "                  samIdx : samtools faidx\n",
        "                    name : reference\n",
        "*******************************\n",
        "module scope: p_celeraassembler\n",
        "*******************************\n",
        "            ovlErrorRate : 0.06\n",
        "              genomeSize : 5000000\n",
        "              genFrgFile : True\n",
        "                 merSize : 14\n",
        "        defaultFrgMinLen : 500\n",
        "               xCoverage : 30\n",
        "             libraryName : pacbioReads\n",
        "                   runCA : False\n",
        "              moduleName : P_CeleraAssembler\n",
        "             maxSlotPerc : 1\n",
        "             description : This module wraps the Celera Assembler v7.0\n",
        "                 castats : False\n",
        "               ovlMinLen : 40\n",
        "                 gzipGff : False\n",
        "            asmWatchTime : 600\n",
        "********************\n",
        "module scope: global\n",
        "********************\n",
        "                 version : 2\n",
        "                   fetch : common/protocols/preprocessing/Fetch.1.xml\n",
        "                   state : active\n",
        "                 control : common/protocols/control/NoControl.1.xml\n",
        "                assembly : common/protocols/assembly/PreAssemblerHGAP.2.xml\n",
        "               consensus : common/protocols/consensus/AssemblyPolishing.1.xml\n",
        "             description : DAGCON-based HGAP (Hierarchical Genome Assembly Process) performs high quality de novo assembly using a single PacBio library prep. HGAP consists of pre-assembly, de novo assembly with Celera Assembler, and assembly polishing with Quiver. This workflow uses a much faster DAG-based consensus approach for pre-assembly.\n",
        "               filtering : common/protocols/filtering/PreAssemblerSFilter.1.xml\n",
        "                 mapping : common/protocols/mapping/BLASR.1.xml\n",
        "            otfReference : reference\n",
        "       referenceUploader : common/protocols/referenceuploader/ReferenceUploaderHGAP.1.xml\n",
        "           deferRefCheck : True\n",
        "                    name : RS_HGAP_Assembly\n",
        "**********************\n",
        "module scope: p_filter\n",
        "**********************\n",
        "              moduleName : P_Filter\n",
        "        minSubReadLength : 500\n",
        "             description : Filter reads for use in the pre-assembly step of HGAP, the hierarchical genome assembly process.\n",
        "               minLength : 100\n",
        "               readScore : 0.80\n",
        "              moduleName : P_FilterReports\n",
        "*********************\n",
        "module scope: p_fetch\n",
        "*********************\n",
        "              moduleName : P_Fetch\n",
        "             description : Sets up inputs\n",
        "*****************************\n",
        "module scope: p_filterreports\n",
        "*****************************\n",
        "              moduleName : P_FilterReports\n",
        "**********************************\n",
        "module scope: p_preassemblerdagcon\n",
        "**********************************\n",
        "       minLongReadLength : 6000\n",
        "              totalBestn : 24\n",
        "              moduleName : P_PreAssemblerDagcon\n",
        "              splitBestn : 10\n",
        "            targetChunks : 6\n",
        "     computeLengthCutoff : True\n",
        "               blasrOpts : -noSplitSubreads -minReadLength 200 -maxScore -1000 -maxLCPLength 16\n",
        "                   title : Using DAG-based consensus algorithm, pre-assemble long reads as the first step of the Hierarchical Genome Assembly process (HGAP). Version 2 is a stepping stone for scaling to much larger genomes.\n",
        "*********************************\n",
        "module scope: p_assemblypolishing\n",
        "*********************************\n",
        "              moduleName : P_AssemblyPolishing\n",
        "             description : Polish a pure-PacBio assembly for maximum accuracy using the Quiver algorithm.\n",
        "       enableMapQVFilter : True\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "required_task_klasses = [i.file_klass for i in pbd.find_required_input_files(abstract_g)]\n",
      "inputs_d = {i.__class__:i.path for i in pbsmrtpipe.mock.generate_mock_files_from_klasses(required_task_klasses, workflow_settings.job_output_dir)}\n",
      "# Overwrite with real reference\n",
      "inputs_d[ReferenceEntryType] = pbsmrtpipe.mock.get_reference_dir()\n",
      "inputs_d"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 7,
       "text": [
        "{pbsmrtpipe.resource_types.ReferenceEntryType: '/Users/mkocher/MOCK_SEYMOUR_HOME/common/references/ecoli_k12_MG1655',\n",
        " pbsmrtpipe.resource_types.FilteredSubreadsFasta: '/tmp/mock_job/filtered_subreads.fasta',\n",
        " pbsmrtpipe.resource_types.PacBioToCASem: '/tmp/mock_job/pacBioToCA.sem',\n",
        " pbsmrtpipe.resource_types.PacBioToCASemFrgLoaded: '/tmp/mock_job/pacBioToCAFrgLoad.sem'}"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_klasses"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "[pbsmrtpipe.task_modules.preassembler_dagcon.FilterLongReadsByLengthTask,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.HgapAlignForCorrectionTask,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.HgapFilterM4Task,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.HgapCorrectionTask,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.PreAssemblerJsonReportTask,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.RunHgapReferenceUploader,\n",
        " pbsmrtpipe.task_modules.preassembler_dagcon.RunHgapCaTask,\n",
        " pbsmrtpipe.task_modules.celera_assembler.WriteRunCASpecTask,\n",
        " pbsmrtpipe.task_modules.celera_assembler.GenerateFrgFileTask]"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t = task_klasses[0]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print tb.pretty_print_task_klass_options([t])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-----------------------------------------------------------------\n",
        "PB module pbsmrtpipe.task_modules.preassembler_dagcon noptions: 9\n",
        "-----------------------------------------------------------------\n",
        "hgapPreTmp            StringOption     True  /FIXME/TMP/dir\n",
        "minLongReadLength     IntRangeOption   False None\n",
        "targetChunks          IntOption        True  1\n",
        "blasrOpts             StringOption     True  \n",
        "blasrChunks           IntRangeOption   True  16\n",
        "genomeSize            IntOption        True  5000000\n",
        "splitBestn            IntOption        True  4\n",
        "totalBestn            IntOption        True  24\n",
        "computeLengthCutoff   BoolOption       True  False\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pbl.resolve_task_options(t, workflow_settings._settings)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "{}"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t._OPTIONS"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "{'blasrChunks': <IntRangeOption name:blasrChunks type:<type 'int'> default_value:16 optional? True >,\n",
        " 'blasrOpts': <StringOption name:blasrOpts type:<type 'str'> default_value: optional? True >,\n",
        " 'computeLengthCutoff': <BoolOption name:computeLengthCutoff type:<type 'bool'> default_value:False optional? True >,\n",
        " 'genomeSize': <IntOption name:genomeSize type:<type 'int'> default_value:5000000 optional? True >,\n",
        " 'hgapPreChunks': <IntOption name:targetChunks type:<type 'int'> default_value:1 optional? True >,\n",
        " 'hgapPreTmp': <StringOption name:hgapPreTmp type:<type 'str'> default_value:/FIXME/TMP/dir optional? True >,\n",
        " 'minLongReadLength': <IntRangeOption name:minLongReadLength type:<type 'int'> default_value:None optional? False >,\n",
        " 'splitBestn': <IntOption name:splitBestn type:<type 'int'> default_value:4 optional? True >,\n",
        " 'totalBestn': <IntOption name:totalBestn type:<type 'int'> default_value:24 optional? True >}"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t.INPUTS, t.OUTPUTS"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "((pbsmrtpipe.resource_types.FilteredSubreadsFasta,),\n",
        " (pbsmrtpipe.resource_types.FilteredLongReadsFasta,))"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "inputs = [pbrt.FilteredSubreadsFasta('/path/to/subreads.fasta')]\n",
      "outputs = [pbrt.FilteredLongReadsFasta('/path/to/filtered_longreads.fasta')]\n",
      "task = t(inputs, outputs, options=dict(minLongReadLength=2))\n",
      "task_io = t.from_io_options(inputs, outputs, options=dict(minLongReadLength=3))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task.resolved_options"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "{'blasrChunks': 16,\n",
        " 'blasrOpts': '',\n",
        " 'computeLengthCutoff': False,\n",
        " 'genomeSize': 5000000,\n",
        " 'hgapPreTmp': '/FIXME/TMP/dir',\n",
        " 'minLongReadLength': 2,\n",
        " 'splitBestn': 4,\n",
        " 'targetChunks': 1,\n",
        " 'totalBestn': 24}"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task.workflow_settings"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "{'blasrChunks': 16,\n",
        " 'blasrOpts': '',\n",
        " 'computeLengthCutoff': False,\n",
        " 'genomeSize': 5000000,\n",
        " 'hgapPreChunks': 1,\n",
        " 'hgapPreTmp': '/FIXME/TMP/dir',\n",
        " 'minLongReadLength': 2,\n",
        " 'splitBestn': 4,\n",
        " 'totalBestn': 24}"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task.to_cmd()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 17,
       "text": [
        "['l=',\n",
        " 'l=${l:-2}',\n",
        " 'echo $l > /path/to/filtered_longreads.fasta.cutoff',\n",
        " 'set -o pipefail',\n",
        " \"fastalength /path/to/subreads.fasta | awk -v len=$l '($1 < len ){ print $2 } '| fastaremove /path/to/subreads.fasta stdin > /path/to/filtered_longreads.fasta \",\n",
        " \"if [ ! -s /path/to/filtered_longreads.fasta ]; then echo 'Empty long reads'; exit 1; fi\"]"
       ]
      }
     ],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task._raw_supplied_options"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 18,
       "text": [
        "{'minLongReadLength': 2}"
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_io.minLongReadLength"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 19,
       "text": [
        "3"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_io.to_cmd()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "['l=',\n",
        " 'l=${l:-3}',\n",
        " 'echo $l > /path/to/filtered_longreads.fasta.cutoff',\n",
        " 'set -o pipefail',\n",
        " \"fastalength /path/to/subreads.fasta | awk -v len=$l '($1 < len ){ print $2 } '| fastaremove /path/to/subreads.fasta stdin > /path/to/filtered_longreads.fasta \",\n",
        " \"if [ ! -s /path/to/filtered_longreads.fasta ]; then echo 'Empty long reads'; exit 1; fi\"]"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "options = dict(minLongLength=2)\n",
      "print options.keys()\n",
      "print t._OPTIONS.keys()\n",
      "for internal_name, o in t._OPTIONS.iteritems():\n",
      "    if internal_name in options:\n",
      "        value = options[internal_name]\n",
      "        print internal_name, value\n",
      "    else:\n",
      "        if o.optional:\n",
      "            # use default\n",
      "            options[internal_name] = o.default_value\n",
      "        else:\n",
      "            print 'Unable to find {n} in {o}'.format(n=internal_name, o=options.keys())\n",
      "            print 'Bad', internal_name, o.name, options.keys()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "['minLongLength']\n",
        "['hgapPreTmp', 'minLongReadLength', 'totalBestn', 'blasrOpts', 'blasrChunks', 'genomeSize', 'splitBestn', 'hgapPreChunks', 'computeLengthCutoff']\n",
        "Unable to find minLongReadLength in ['hgapPreTmp', 'minLongLength']\n",
        "Bad minLongReadLength minLongReadLength ['hgapPreTmp', 'minLongLength']\n"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 22,
       "text": [
        "pbsmrtpipe.task_modules.preassembler_dagcon.FilterLongReadsByLengthTask"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t._OPTIONS.keys()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 23,
       "text": [
        "['hgapPreTmp',\n",
        " 'minLongReadLength',\n",
        " 'totalBestn',\n",
        " 'blasrOpts',\n",
        " 'blasrChunks',\n",
        " 'genomeSize',\n",
        " 'splitBestn',\n",
        " 'hgapPreChunks',\n",
        " 'computeLengthCutoff']"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t._OPTIONS['minLongReadLength'].value(2)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 26,
       "text": [
        "2"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t._OPTIONS"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 27,
       "text": [
        "{'blasrChunks': <IntRangeOption name:blasrChunks type:<type 'int'> default_value:16 optional? True >,\n",
        " 'blasrOpts': <StringOption name:blasrOpts type:<type 'str'> default_value: optional? True >,\n",
        " 'computeLengthCutoff': <BoolOption name:computeLengthCutoff type:<type 'bool'> default_value:False optional? True >,\n",
        " 'genomeSize': <IntOption name:genomeSize type:<type 'int'> default_value:5000000 optional? True >,\n",
        " 'hgapPreChunks': <IntOption name:targetChunks type:<type 'int'> default_value:1 optional? True >,\n",
        " 'hgapPreTmp': <StringOption name:hgapPreTmp type:<type 'str'> default_value:/FIXME/TMP/dir optional? True >,\n",
        " 'minLongReadLength': <IntRangeOption name:minLongReadLength type:<type 'int'> default_value:None optional? False >,\n",
        " 'splitBestn': <IntOption name:splitBestn type:<type 'int'> default_value:4 optional? True >,\n",
        " 'totalBestn': <IntOption name:totalBestn type:<type 'int'> default_value:24 optional? True >}"
       ]
      }
     ],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_io.resolved_options"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 28,
       "text": [
        "{'blasrChunks': 16,\n",
        " 'blasrOpts': '',\n",
        " 'computeLengthCutoff': False,\n",
        " 'genomeSize': 5000000,\n",
        " 'hgapPreTmp': '/FIXME/TMP/dir',\n",
        " 'minLongReadLength': 3,\n",
        " 'splitBestn': 4,\n",
        " 'targetChunks': 1,\n",
        " 'totalBestn': 24}"
       ]
      }
     ],
     "prompt_number": 28
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print task_io.minLongReadLength"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "3\n"
       ]
      }
     ],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pbl.resolve_task_options(t, workflow_settings._settings)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 32,
       "text": [
        "{}"
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_io.workflow_settings"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 30,
       "text": [
        "{'blasrChunks': 16,\n",
        " 'blasrOpts': '',\n",
        " 'computeLengthCutoff': False,\n",
        " 'genomeSize': 5000000,\n",
        " 'hgapPreChunks': 1,\n",
        " 'hgapPreTmp': '/FIXME/TMP/dir',\n",
        " 'minLongReadLength': 3,\n",
        " 'splitBestn': 4,\n",
        " 'totalBestn': 24}"
       ]
      }
     ],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "resolved_graph = pbd.build_resolved_graph(abstract_g, inputs_d, workflow_settings)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "ValueError",
       "evalue": "Required option 'minLongReadLength' (XML/IO name 'minLongReadLength') of type <type 'int'> not supplied in task FilterLongReadsByLengthTask in pbsmrtpipe.task_modules.preassembler_dagcon. \nValid options names ['hgapPreTmp', 'minLongReadLength', 'totalBestn', 'blasrOpts', 'blasrChunks', 'genomeSize', 'splitBestn', 'hgapPreChunks', 'computeLengthCutoff'].\n Supplied options {}",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-31-839f15aa1205>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mresolved_graph\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpbd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuild_resolved_graph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mabstract_g\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minputs_d\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworkflow_settings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;32m/Users/mkocher/.virtualenvs/dev_pbsmrtpipe/lib/python2.7/site-packages/pbsmrtpipe/dag.pyc\u001b[0m in \u001b[0;36mbuild_resolved_graph\u001b[0;34m(g, inputs_d, workflow_settings)\u001b[0m\n\u001b[1;32m    350\u001b[0m                                           \u001b[0mtmp_files\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtmp_files\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    351\u001b[0m                                           \u001b[0mtmp_dirs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtmp_dirs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 352\u001b[0;31m                                           options=resolved_options)\n\u001b[0m\u001b[1;32m    353\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    354\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/Users/mkocher/.virtualenvs/dev_pbsmrtpipe/lib/python2.7/site-packages/pbsmrtpipe/model.pyc\u001b[0m in \u001b[0;36mfrom_io_options\u001b[0;34m(cls, inputs, outputs, nproc, tmp_dirs, tmp_files, options)\u001b[0m\n\u001b[1;32m    639\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    640\u001b[0m         return cls(inputs, outputs, nproc=nproc, tmp_files=tmp_files,\n\u001b[0;32m--> 641\u001b[0;31m                    tmp_dirs=tmp_dirs, options=internal_options)\n\u001b[0m\u001b[1;32m    642\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    643\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__getattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/Users/mkocher/.virtualenvs/dev_pbsmrtpipe/lib/python2.7/site-packages/pbsmrtpipe/model.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, inputs, outputs, nproc, tmp_dirs, tmp_files, options)\u001b[0m\n\u001b[1;32m    577\u001b[0m                               \u001b[0ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_OPTIONS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    578\u001b[0m                               o=options)\n\u001b[0;32m--> 579\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Required option '{n}' (XML/IO name '{i}') of type {t} not supplied in task {k} in {m}. \\nValid options names {s}.\\n Supplied options {o}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0m_d\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    580\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    581\u001b[0m         \u001b[0;31m# Detail Inputs/Outputs Validation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;31mValueError\u001b[0m: Required option 'minLongReadLength' (XML/IO name 'minLongReadLength') of type <type 'int'> not supplied in task FilterLongReadsByLengthTask in pbsmrtpipe.task_modules.preassembler_dagcon. \nValid options names ['hgapPreTmp', 'minLongReadLength', 'totalBestn', 'blasrOpts', 'blasrChunks', 'genomeSize', 'splitBestn', 'hgapPreChunks', 'computeLengthCutoff'].\n Supplied options {}"
       ]
      }
     ],
     "prompt_number": 31
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tb.display_networkx_graph(resolved_graph)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'resolved_graph' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-28-c3268ba73cb9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisplay_networkx_graph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresolved_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;31mNameError\u001b[0m: name 'resolved_graph' is not defined"
       ]
      }
     ],
     "prompt_number": 28
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tns = resolved_graph.task_nodes()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "task_klasses"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 7,
       "text": [
        "[pbsmrtpipe.task_modules.fetch.ToFofnTask,\n",
        " pbsmrtpipe.task_modules.fetch.OverViewTask,\n",
        " pbsmrtpipe.task_modules.fetch.GetChemistryTask,\n",
        " pbsmrtpipe.task_modules.fetch.AdapterReportTask,\n",
        " pbsmrtpipe.task_modules.filter.FilterTask,\n",
        " pbsmrtpipe.task_modules.filter.FilterSubreadsTask,\n",
        " pbsmrtpipe.task_modules.filter.FilterSubreadSummaryTask]"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tf = task_klasses[4]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fo = tf._OPTIONS['filters']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "workflow_settings._settings"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "{'global.noise.data': '-77.27,0.08654,0.00121',\n",
        " 'global.reference': '/mnt/secondary-siv/references/ecoli_k12_MG1655',\n",
        " 'global.use_subreads': 'True',\n",
        " 'p_filter.filters': 'MinRL=50,MinReadScore=0.65',\n",
        " 'p_filter.trim': 'True',\n",
        " 'p_mapping.concordant': 'True',\n",
        " 'p_mapping.gff2Bed': 'True',\n",
        " 'p_mapping.loadPulsesOpts': 'bymetric',\n",
        " 'p_mapping.maxDivergence': '30',\n",
        " 'p_mapping.maxHits': '10',\n",
        " 'p_mapping.minAnchorSize': '12',\n",
        " 'p_mapping.pbalign_opts': '--seed=1 --minAccuracy=0.75 --minLength=50 --algorithmOptions=\"-useQuality\"',\n",
        " 'p_mapping.placeRepeatsRandomly': 'True',\n",
        " 'p_mapping.pulseMetrics': 'DeletionQV,IPD,InsertionQV,PulseWidth,QualityValue,MergeQV,SubstitutionQV,DeletionTag',\n",
        " 'p_mapping.samBam': 'True'}"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fo.value('MinRL=50,MinReadScore=0.65')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "'MinRL=50,MinReadScore=0.65'"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tf.INPUTS"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "(pbsmrtpipe.resource_types.MovieFofn,)"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tf.OUTPUTS"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "(pbsmrtpipe.resource_types.FilteredRegionFofn,\n",
        " pbsmrtpipe.resource_types.FilterSummaryCsv)"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "movie_fofn = pbrt.MovieFofn(\"/path/to/movie.fofn\")\n",
      "rfofn = pbrt.FilteredRegionFofn(\"/path/to/filtered_region.fofn\")\n",
      "csv = pbrt.FilterSummaryCsv(\"/path/to/filtered_subreads.csv\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task = tf([movie_fofn], [rfofn, csv], 1, trim='True', filters='MinRL=50,MinReadScore=0.65', minSubReadLength=50, readScore=0.75,use_subreads='True')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task.workflow_settings"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "{'artifact_score': None,\n",
        " 'filters': 'MinRL=50,MinReadScore=0.65',\n",
        " 'maxLength': None,\n",
        " 'maxSubReadLength': None,\n",
        " 'minLength': None,\n",
        " 'minSNR': None,\n",
        " 'minSubReadLength': 50,\n",
        " 'readScore': 0.75,\n",
        " 'trim': True,\n",
        " 'use_ccs': False,\n",
        " 'use_subreads': True,\n",
        " 'whiteList': None}"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print filter_task.trim\n",
      "print filter_task.use_subreads\n",
      "print filter_task.filters\n",
      "print getattr(filter_task, 'minSubReadLength')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "True\n",
        "True\n",
        "MinRL=50,MinReadScore=0.65\n",
        "50\n"
       ]
      }
     ],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task.to_cmd()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 18,
       "text": [
        "['mkdir -p /path/to/filtered_regions',\n",
        " \"filter_plsh5.py --debug --filter='MinRL=50,MinReadScore=0.65,MinReadScore=0.75,MinSRL=50' --trim='True' --outputDir=/path/to/FIXME --outputSummary=/path/to/filtered_subreads.csv --outputFofn=/path/to/FIXME2 /path/to/movie.fofn\"]"
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print filter_task.trim"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "True\n"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task._OPTIONS['trim'].from_string('True')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "True"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print getattr(filter_task, 'trim')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "True\n"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task.resolved_options"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 22,
       "text": [
        "{}"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task.supplied_options"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 23,
       "text": [
        "{'filters': 'MinRL=50,MinReadScore=0.65',\n",
        " 'minSubReadLength': 50,\n",
        " 'readScore': 0.75,\n",
        " 'trim': 'True',\n",
        " 'use_subreads': 'True'}"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_task.workflow_settings"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 24,
       "text": [
        "{'artifact_score': None,\n",
        " 'filters': 'MinRL=50,MinReadScore=0.65',\n",
        " 'maxLength': None,\n",
        " 'maxSubReadLength': None,\n",
        " 'minLength': None,\n",
        " 'minSNR': None,\n",
        " 'minSubReadLength': 50,\n",
        " 'readScore': 0.75,\n",
        " 'trim': True,\n",
        " 'use_ccs': False,\n",
        " 'use_subreads': True,\n",
        " 'whiteList': None}"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "o = pbsmrtpipe.model.BoolOption('my_bool', default_value=True, optional=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "o.from_string('True')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 26,
       "text": [
        "True"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}